Technical / editorial: Document the text-box-* properties (#37738)
mdn/content: files/en-us/web/css/text-box-edge/index.md

---
title: text-box-edge
slug: Web/CSS/text-box-edge
page-type: css-property
status:
  - experimental
browser-compat: css.properties.text-box-edge
---

{{CSSRef}}{{seecompattable}}

> The **`text-box-edge`** [CSS](/en-US/docs/Web/CSS) property specifies an amount of [leading](/en-US/docs/Glossary/Leading) to trim from text content.

Reviewer: ```suggestion
The **`text-box-edge`** [CSS](/en-US/docs/Web/CSS) property specifies the trim direction when trimming text content using the {{cssxref("text-box-trim")}} property or {{cssxref("text-box")}} shorthand. It has no effect if `text-box-trim` is set or defaults to none.
```
Not 100% sure if what i wrote is accurrate, but it doesn't specify an amount or impact leading,

Author: Hrm, this is getting more confusing by the minute. I've just done a test that sets these properties on a multiline text box, and they have no effect on the spacing between the lines of text. Just the spacing at the start and/or edge of the overall text box.

So, I think the leading might be trimmed from the start and end of the text box (depending on what `text-box-edge` value is chosen), but not in between the lines. 

But I'm not sure. @argyleink, do you have an answer here, or any thoughts?

> Leading differs across fonts, making consistent typesetting historically challenging on the web. `text-box-edge` — along with its counterpart property {{cssxref("text-box-trim")}}, which specifies which edge(s) to trim leading from — makes consistent typesetting easier to achieve.

Reviewer: ```suggestion
Text height differs between fonts, making consistent typesetting historically challenging on the web. The `text-box-edge` — along with its counterpart property {{cssxref("text-box-trim")}} — makes consistent typesetting easier to achieve by enabling trimming of extra spacing above or below text. With the {{cssxref("text-edge")}} properties, we can set the over and under edges for inline boxes. This enables vertically aligning text regardless of the font file's baseline.
```

Author: I'm leaving this one until we can clear up the earlier confusion. Leading must be involved — the relevant spec section is titled [Trimming Leading Over/Under Text](https://drafts.csswg.org/css-inline-3/#leading-trim) — but after your review, I'm just confused as to exactly how (although I suspect it is removed from just the start and end of the entire text box).

Again, @argyleink, I'd love to hear your thoughts on this.

Reviewer 2: Technically speaking, the trimming is applied to the block container, not to inline box nor to text box, using the metrics of the font of the block container. [This jsbin](https://jsbin.com/pederax/edit?html,output) shows how each box is modified and not modified. Does this help understanding?

> Leading differs across fonts, making consistent typesetting historically challenging on the web. `text-box-edge` — along with its counterpart property {{cssxref("text-box-trim")}}, which specifies which edge(s) to trim leading from — makes consistent typesetting easier to achieve.

Author: updated

> [!NOTE]

> > There is also a shorthand property — {{cssxref("text-box")}}, which can be used to specify the `text-box-edge` and `text-box-trim` values in a single declaration.

Reviewer: ```suggestion
> The {{cssxref("text-box")}} shorthand property can be used to specify the `text-box-edge` and `text-box-trim` values in a single declaration.
```
if we go with my suggested edit that includes the shorthand

Author: It's a good edit regardless; updated.

## Syntax

```css
/* Single keyword */
text-box-edge: auto;
text-box-edge: text;

> /* Two keywords */

Author: Yup, good point. I've gone with "Two <text-edge> values" as I felt it was clearer.

text-box-edge: text text;
text-box-edge: text alphabetic;
text-box-edge: cap alphabetic;
text-box-edge: ex text;

/* Global values */
text-box-edge: inherit;
text-box-edge: initial;
text-box-edge: revert;
text-box-edge: revert-layer;
text-box-edge: unset;
```

### Value

> The `text-box-edge` property value is specified as the keyword `auto` or a {{cssxref("&lt;text-edge&gt;")}} value, which may contain one or two separate keywords representing over and under edge positions to trim the font leading to.

Author: Updated to use this structure.

- `auto`

>   - : Equivalent to the `text-edge` value `text`.

Reviewer 2: I think this is intentional, but just wanted to double check. The `auto` is equivalent to `text` only if the browser hasn't implemented the `line-fit-edge` property. Wondering if this is nice to mention, or is this unnecessary complexity?

Author: I think we should leave this as-is for now, as it describes the current implementation. We should try to avoid complicating the description if we can.

When browsers start to implement `line-fit-edge`, we should revisit the description then.

Of course, if some browsers support it and others don't, we'll have to fork the description.

What is the behavior if `line-fit-edge` _is_ implemented?

Reviewer 2: sgtm, thanks for the confirmation.

> What is the behavior if `line-fit-edge` is implemented?

[The current spec says](https://drafts.csswg.org/css-inline-3/#text-box-edge:~:text=the%20auto%20keyword%20uses%20the%20value%20of%20line%2Dfit%2Dedge%2C%20interpreting%20leading%20(the%20initial%20value)%20as%20text.):
> the `auto` keyword uses the value of `line-fit-edge`, interpreting `leading` (the initial value) as `text`.

Author: Thanks, that makes sense. Yeah, we'll update this when `line-fit-edge` is implemented.

- {{cssxref("&lt;text-edge&gt;")}}
  - : One or two separate keywords representing over and under edge positions to trim the font leading to.
    - If two values are specified, the first value specifies the trimming behavior to apply to the block-start (over) edge of the text, and the second value specifies the trimming behavior to apply to the block-end (under) edge of the text.
      - Valid over edge trimming values: `text` (the default), `cap`, and `ex`.

>       - Valid under edge trimming values: `text` (the default) and `alphabetic`.

Reviewer 2: Can you please remove "(the default)"? There are no longer a "default" value, please see https://github.com/w3c/csswg-drafts/issues/10703.

Author: I've removed both instances of "(the default)" in my next commit, but ... there must be a default if the property is not explicitly set. I did some testing and saw that, in the latest Chrome implementation:

- just setting a single value of `cap`, `ex`, or `alphabetic` is flagged as invalid by the devtools, so my current text is correct.
- the computed value when `text-box-edge` is not set on an element is `auto`. So I've added a line to the `auto` definition on the page saying "This is the default value."

Does that work for you?

Reviewer 2: > there must be a default if the property is not explicitly set

Oh I see, do you mean the "initial"? In the CSS terminologies, the "default" is a bit ambiguous, and I thought you meant it for a default value when only one value is used. Does "initial" work for you? I understand "default" may sound more familiar though.

Author: Yup, that's how we tend to refer to the initial value set by the browser stylesheet. "initial" is a bit more correct as per official CSS terminology, but "default" is a bit easier to understand. Plus it makes the use of "default" more consistent with how we talk about default values in JS and WebAPI ref docs.

At least, that's what I've been doing. @estelle, what are your thoughts on this?

> The specification also defines the `ideographic` and `ideographic-ink` keywords, which are intended to represent regions of [CJK language characters](https://en.wikipedia.org/wiki/CJK_characters). Currently their exact behavior is being debated and they are are not supported by any browser.

Reviewer: ```suggestion
```
can be removed as we link to the data type, which explains all that, and we don't list the enumerated values here.

## Description

> Font files specify an amount of space — termed leading — to be included above and below text, to contain capital letters, ascenders, descenders, etc., and provide spacing between lines. The amount of leading included varies between fonts, making it difficult to consistently set block-level text spacing.

Reviewer: ```suggestion
Font files specify the space included above and below text, to contain capital letters, ascenders, descenders, etc. This in-font spacing ensures lines of text don't overlap by default. The amount of spacing included varies between fonts, making it difficult to consistently set block-level text spacing.
```

> The `text-box-edge` property allows you to trim the leading from the start and/or end edge of the text. It does this by specifying a {{cssxref("&lt;text-edge&gt;")}} value that indicates the over edge and under edge to trim the leading to.

Reviewer: ```suggestion
The `text-box-edge` property allows you to trim the spacing from the start and/or end edge of lines of text by specifying a {{cssxref("&lt;text-edge&gt;")}} value that indicates the over edge and under edge to trim the inline box.
```

Reviewer: i think we should get rid of this "leading" since that is between lines, not within the inline box. I'll stop editing here and give you a chance to marinate / agree / disagree / get a pint / get several pints ;)

Author: I'm leaving this one just for now, until we clear up the confusion.

Which edge(s) to trim space from is specified using the {{cssxref("text-box-trim")}} property. For example, you can choose to trim space from the over edge or the under edge of the text's block container, or both.

These properties make it much easier to control text spacing in the block direction.

## Formal definition

{{cssinfo}}

## Formal syntax

{{csssyntax}}

## Examples

### Basic `text-box-edge` usage

> The most common `text-box-edge` values you'll use for horizontal {{cssxref("writing-mode")}} languages such as English or Arabic are `cap alphabetic` and `ex alphabetic`. The `cap` value trims the over edge of the text element's block container to the top of the capital letters, whereas `ex` trims the over edge to the font's x-height (the top edge of the short lower-case letters). In each case, `alphabetic` trims the under-edge flush with the text baseline.

Reviewer 3: Not sure if this will work reliably for Arabic, actually. It might work incidentally for a number of fonts, but Arabic text is not fundamentally tied to the same metrics as Latin in the same way, e.g. Greek typically is.

Author: OK, noted. So I should just remove "or Arabic" from here?
